optimizer/typeinferer.go

```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
 	case *ast.FuncCastExpr:
 		x.SetType(x.Tp)
 	case *ast.SelectStmt:
-		rf := x.GetResultFields()
-		for _, val := range rf {
-			if val.Column.ID == 0 && val.Expr.GetType() != nil {
-				val.Column.FieldType = *(val.Expr.GetType())
+		v.selectStmt(x)
+	case *ast.ParamMarkerExpr:
+		x.SetType(types.DefaultTypeForValue(x.GetValue()))
+	case *ast.BinaryOperationExpr:
+		v.binaryOperation(x)
+	case *ast.UnaryOperationExpr:
+		v.unaryOperation(x)
+	case *ast.BetweenExpr:
```

Comment:
If we put multiple type in a case, we can't use the real typed `x`, we can only use the `in`,
`in` is `ast.Node`, not `ast.ExprNode`, it doesn't have `SetType` method.